src/rgw/rgw_role.cc

```diff
@@ -82,10 +84,10 @@ int RGWRole::create(bool exclusive)
   struct timeval tv;
   real_clock::to_timeval(t, tv);
 
-  char buf[30];
-  strftime(buf,30,"%Y-%m-%dT%H:%M:%S", std::gmtime(&tv.tv_sec));
-  sprintf(buf + strlen(buf),".%dZ",(int)tv.tv_usec);
-  creation_date.assign(buf, strlen(buf));
+  std::ostringstream oss;
+  oss << std::put_time(std::gmtime(&tv.tv_sec), "%Y-%m-%dT%H:%M:%S");
+  oss << "." << (tv.tv_usec/1000) << "Z";
+  creation_date = oss.str();
 
```

Comment:
C++ time formatting is, unfortunately, a bit of a fiasco so I'm not sure there's a really *good* solution. strftime is probably faster since it has less overhead, not needing to call the entire iostream infrastructure into place.

The one thing is, don't use std::gmtime. It's not guaranteed to be thread safe (and strict interpretation of the C standard forbids it to be thread safe, since all calls to gmtime and localtime are required to return the same pointer). Use gmtime_r. It's not C++ but is in the standard C library. It started out as a GNU extension but has been folded into the SIngle Unix Specification so should be available everywhere.